phy/phy.nim

```diff
@@ -0,0 +1,346 @@
+        error "unknown option: " & key
+    of cmdArgument:
+      if cmd == None:
+        cmd = parseEnum[Command](key)
+      else:
+        # must be the input filename
+        input = key
+    of cmdEnd:
+      unreachable()
+
+  # make sure the command line arguments are sensible:
+  case cmd
+  of None:
+    echo HelpText
+    quit(1)
+  of Eval:
+    if target != langBytecode:
+      error "'e' requires the target language to be '" & $langBytecode & "'"
+  else:
+    discard "nothing special to do"
+
+  if ord(target) > ord(source):
+    error "invalid source and target language pair"
+
+  if input.len == 0:
+    error "a filename must be provided"
+
+  var s = openFileStream(input, fmRead)
+  if gRunner:
+    # skip the test file header, if any
+    if s.readLine() == "discard \"\"\"":
+      while not s.readLine().endsWith("\"\"\""):
+        discard
+    else:
+      s.setPosition(0)
+
+  let text = s.readAll()
+  s.close()
+
+  var
+    env = initVm(1024, 1024 * 1024) # 1MB of memory
+    code: PackedTree[spec.NodeKind]
+    typ: SemType
+
+  if source == langBytecode:
+    # run the assembler on the input lines
+    var
+      a = AssemblerState()
+      lineN = 1
+
+    for line in splitLines(text, false):
+      try:
+        a.process(line, env)
+      except AssemblerError as e:
+        error input & "(" & $lineN & ", 1): " & e.msg
+      inc lineN
+  else:
+    var newSource = source
+
+    # the input is in some higher-level language
+    if source == langSource:
+      # translate to the highest-level IL first
+      (code, typ) = sourceToIL(text)
+      newSource = lang10
+      print(code, newSource)
+    elif gRunner:
+      # in order to reduce visual noise in tests, the ``(Module ...)`` top
+      # level node is added implicitly
+      code = fromSexp[spec.NodeKind](parseSexp("(Module " & text & ")"))
+    else:
+      code = fromSexp[spec.NodeKind](parseSexp(text))
+
+    if target == langBytecode:
+      # compile to L0 code and then translate to bytecode
+      compile(code, newSource, lang0)
+      syntaxCheck(code, lang0)
+      pass0.translate(code, env)
+      # the bytecode is verified later
+    else:
+      compile(code, newSource, target)
+      # make sure the output code is correct:
+      syntaxCheck(code, target)
+
+  if target == langBytecode:
+    # make sure the environment is correct:
+    let errors = validate(env)
+    if errors.len > 0:
+      echo "Validation of the VM environment failed"
+      for it in errors.items:
+        echo "Error: ", it
+      quit(1)
+
+    print(env)
+
+    # handle the eval command:
+    if cmd == Eval:
+      if env.procs.len == 0:
+        error "there's nothing to run"
+
+      if source == langSource:
+        # we have type high-level type information
+        stdout.write run(env, env.procs.high.ProcIndex, typ)
+      else:
+        # we don't have high-level type information
+        stdout.write run(env, env.procs.high.ProcIndex)
+
+main(getExecArgs())
```

Comment:
```suggestion
when isMainModule:
  main(getExecArgs())
```

This would allow importing the module if one wanted. In particular, I think it would mean we could embed the functionality of this module in the runners and incur less overhead?